backend/tracim_backend/lib/core/user.py

```diff
@@ -1121,3 +1123,102 @@ def allowed_to_invite_new_user(self, email: str) -> bool:
+    ) -> UserFollower:
+        # NOTE BS 20201229: sqlalchemy raise database specific error. In addition, error is
+        # raised at commit, so after view execution (so we can't catch it with hapic view
+        # decorators). So check if row exist before try to insert.
+        if (
+            self._session.query(UserFollower)
+            .filter(
+                and_(UserFollower.leader_id == leader_id, UserFollower.follower_id == follower_id)
+            )
+            .count()
+        ):
+            raise UserFollowAlreadyDefined("User follow already defined")
+
+        user_follower = UserFollower(
+            follower_id=follower_id, leader_id=leader_id, created_date=created_date
+        )
+        self._session.add(user_follower)
+
+        if do_save:
+            self._session.flush()
+
+        return user_follower
+
+    def delete_user_following(self, follower_id: int, leader_id: int, do_save: bool = True) -> None:
+        try:
+            user_follow = (
+                self._session.query(UserFollower)
+                .filter(
+                    and_(
+                        UserFollower.leader_id == leader_id, UserFollower.follower_id == follower_id
+                    )
+                )
+                .one()
+            )
+            self._session.delete(user_follow)
+        except NoResultFound:
+            raise NotFound("User following does not exist")
+
+        if do_save:
+            self._session.flush()
+
+    def get_paginated_user_following_for_user(
+        self,
+        user_id: int,
+        count: typing.Optional[int] = DEFAULT_NB_ITEM_PAGINATION,
+        page_token: typing.Optional[int] = None,
+        filter_user_id: typing.Optional[int] = None,
+    ) -> Page:
+        query = (
+            self._session.query(UserFollower.leader_id.label("user_id"))
+            .filter(UserFollower.follower_id == user_id)
+            .order_by(UserFollower.leader_id)
+        )
+        if filter_user_id is not None:
+            query = query.filter(UserFollower.leader_id == filter_user_id)
+        return get_page(query, per_page=count, page=page_token or False)
+
+    def get_paginated_user_followers_for_user(
+        self,
+        user_id: int,
+        count: typing.Optional[int] = DEFAULT_NB_ITEM_PAGINATION,
+        page_token: typing.Optional[int] = None,
+        filter_user_id: typing.Optional[int] = None,
+    ) -> Page:
+        query = (
+            self._session.query(UserFollower.follower_id.label("user_id"))
+            .filter(UserFollower.leader_id == user_id)
+            .order_by(UserFollower.follower_id)
+        )
+        if filter_user_id is not None:
+            query = query.filter(UserFollower.follower_id == filter_user_id)
+        return get_page(query, per_page=count, page=page_token or False)
+
+    def get_user_follow_count(
+        self, follower_id: typing.Optional[int] = None, leader_id: typing.Optional[int] = None
+    ) -> int:
```

Comment:
I'd rather have two methods (`get_followers_count` + `get_leaders_count`), or have a signature like
```python
class Relationship(enum.Enum):
    FOLLOWER = 1
    LEADER = 2
def get_user_relationship_count(self, user_id: int, relationship: Relationship) -> int:
```
As either way will make clear which parameters are needed only by reading their signatures.

Reply:
ok, i choose `get_followers_count` + `get_leaders_count`. done in next commit